spec/lib/msf/debug_spec.rb

```diff
@@ -0,0 +1,864 @@
+
+RSpec.describe Msf::Ui::Debug do
+  let(:file_fixtures_path) { File.join(Msf::Config.install_root, 'spec', 'file_fixtures') }
+
+  it 'correctly parses an error log' do
+    allow(::Msf::Config).to receive(:log_directory).and_return(File.join(file_fixtures_path, 'debug', 'error_logs', 'basic'))
+
+    error_log_output = <<~LOG
+      ##  %grnErrors%clr
+
+      The following errors occurred before the issue occurred:
+      <details>
+      <summary>Collapse</summary>
+
+      ```
+      [00/00/0000 00:00:00] [e(0)] core: [-] Error 1
+
+      [11/11/1111 11:11:11] [e(0)] core: [-] Error 2
+      Call stack:
+      Stack_Trace
+      stack trace
+      STACK-TRACE
+
+      [22/22/2222 22:22:22] [e(0)] core: [-] Error 3
+      ```
+
+      </details>
+
+
+    LOG
+
+    expect(subject.errors).to eql(error_log_output)
+  end
+
+  it 'correctly parses an error log file larger than the log line total' do
+    allow(::Msf::Config).to receive(:log_directory).and_return(File.join(file_fixtures_path, 'debug', 'error_logs', 'long'))
+
+    logs = ''
+    digits = 11..20
+
+    digits.each do |d|
+      logs += "[00/00/0000 00:00:00] [e(0)] core: [-] Error #{d}\n\n"
+    end
+
+    error_log_output = <<~LOG
+      ##  %grnErrors%clr
+
+      The following errors occurred before the issue occurred:
+      <details>
+      <summary>Collapse</summary>
+
+      ```
+      #{logs.strip}
+      ```
+
+      </details>
+
+
+    LOG
+
+    expect(subject.errors).to eql(error_log_output)
+  end
+
+  it 'correctly parses an empty error log file' do
+    allow(::Msf::Config).to receive(:log_directory).and_return(File.join(file_fixtures_path, 'debug', 'error_logs', 'empty'))
+
+    error_log_output = <<~EMPTY
+      ##  %grnErrors%clr
+
+      The following errors occurred before the issue occurred:
+      <details>
+      <summary>Collapse</summary>
+
+      ```
+      The error log file was empty
+      ```
+
+      </details>
+
+
+    EMPTY
+
+    expect(subject.errors).to eql(error_log_output)
+  end
+
+  it 'correctly retrieves and parses a command history shorter than the command total' do
+    stub_const('Readline::HISTORY', Array.new(4) { |i| "Command #{i + 1}" })
+
+    driver = instance_double(
+      'driver',
+      hist_last_saved: 0
+    )
+
+    stub_const("Msf::Ui::Debug::COMMAND_HISTORY_TOTAL", 10)
+    expect(driver).to receive(:hist_last_saved).exactly(2).times
+
+    history_output = <<~E_LOG
+      ##  %grnHistory%clr
+
+      The following commands were ran during the session and before this issue occurred:
+      <details>
+      <summary>Collapse</summary>
+
+      ```
+      0      Command 1
+      1      Command 2
+      2      Command 3
+      3      Command 4
+      ```
+
+      </details>
+
+
+    E_LOG
+
+    expect(subject.history(driver)).to eql(history_output)
+  end
+
+  it 'correctly retrieves and parses a command history equal in length to the command total' do
+    driver = instance_double(
+      'driver',
```

Comment:
Note that `instance_double` works best with a reference to a class, rather than a string 👀 

In this scenario you lose rspec's ability to detect invalid methods if you use a string rather than a class